# Description: Store the cache stats to disk immediately when a new cache is
#  created. As this normally only happens on a clean shutdown or every 5
#  minutes, the cache stats will never be stored if the application exits
#  uncleanly within 5 minutes. On a subsequent startup, disk_cache::VerifyStats
#  will fail the signature check, causing the cache to be recreated
# Author: Chris Coulson <chris.coulson@canonical.com>

diff --git a/net/disk_cache/blockfile/backend_impl.cc b/net/disk_cache/blockfile/backend_impl.cc
--- a/net/disk_cache/blockfile/backend_impl.cc
+++ b/net/disk_cache/blockfile/backend_impl.cc
@@ -1395,17 +1395,26 @@ bool BackendImpl::InitStats() {
     FileType file_type = Addr::RequiredFileType(size);
     DCHECK_NE(file_type, EXTERNAL);
     int num_blocks = Addr::RequiredBlocks(size, file_type);
 
     if (!CreateBlock(file_type, num_blocks, &address))
       return false;
 
     data_->header.stats = address.value();
-    return stats_.Init(NULL, 0, address);
+    if (!stats_.Init(NULL, 0, address))
+      return false;
+
+    // Stats are only saved to disk on shutdown or every 5 minutes. If we don't
+    // do this now, then exiting uncleanly within 5 minutes if the cache is
+    // newly created will result in it failing to load next time (the signature
+    // check in VerifyStats() will fail)
+    StoreStats();
+
+    return true;
   }
 
   if (!address.is_block_file()) {
     NOTREACHED();
     return false;
   }
 
   // Load the required data.
